### PR TITLE
[mlir] Fix file leak in SeralizeToHsaco

### PR DESCRIPTION
**Pull request by krzysz00**
SerializetToHsaco, as currently implemented, leaks the file descriptor of the .hsaco temporary file, which causes issues in long-running parallel compilation setups.

In addition, move to an LLVM mutex for the linking stage rather than std::muterx to match upstream better.

Fixes
https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1151

### CODE REVIEW

**Review by jerryyin — APPROVED** [review]
I think it is useful to know:
1. How do you figure out it is hsaco file descriptor that's not closed? Through strace?
2. How do you think we can prevent this from happening again? Can we run a single lowering with strace and make a text match with keyword like "*leak*"

**Comment by krzysz00** [comment]
@jerryyin 
As to 1., here's how the debug process went
```bash
```
strace -ff -e trace=%file,%desc -o calls env MIGRAPHX_ENABLE_MLIR=1 ./bin/test_verify test_lstm_reverse_last
cat calls.* > calls
rm calls.*
less calls
```
then skipping to the end (which was an MLIR process or had one close-by) to look at the patterns of `openat`s and `read`s and `close`s.
2. There isn't really an automated way to regression test this that I'm aware of, especially since we're running a bunch of stuff in the same process so it's hard to tell the difference between leaking something and just not having gotten around to closing it yet since you still need it.

File: external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp [hunk 1]
```diff
@@ -448,13 +452,14 @@ SerializeToHsacoPass::createHsaco(const SmallVectorImpl<char> &isaBinary) {
   }
 
   // Load the HSA code object.
-  auto hsacoFile = openInputFile(tempHsacoFilename);
+  auto hsacoFile =
+      llvm::MemoryBuffer::getFile(tempHsacoFilename, /*IsText=*/false);
```
Comment (sjw36) [inline external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:456]: Good, should have a runtime improvement also!

Comment (krzysz00) [reply external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:456]: I don't think there'll be a runtime improvement, we copy this binary out to a vector anyway

**Pull request merged**